File: pna/sparse/F/pollutionModel.f90
Program PollutionModel

  use csrmatrix

  Implicit None

  integer, parameter :: M = 7   ! grid resolution

  integer, parameter      :: maxIter = 5000    ! Maximum iterations of BiCGSTAB 
  integer :: usedIter
  real(kind=8), parameter :: tol     = 1.0d-6  ! Relative BiCGSTAB tolerance

  ! Boundary conditions, -- see exercise sheet for notation
  real(kind=8), parameter :: y1 = 0.4d0
  real(kind=8), parameter :: y2 = 0.6d0
  integer, parameter :: p = 2

  ! k normalises the height of the chimney to 1
  real(kind=8), parameter :: k=(2.0d0/(y2-y1))**(2*p)


  ! indices
  integer :: index   ! index of current element
  integer :: i       ! current i coordinate
  integer :: j       ! current j coordinate

  real(kind=8) :: x, y  ! x and y data at current coordinate

  ! temporary store for coefficient data
  real(kind=8) :: coeffC, coeffN, coeffE, coeffS, coeffW

  ! local indices on interior mesh
  integer :: RC, RN, RE, RS, RW

  ! The linear system:
  integer, parameter :: noUnknowns = M*M
  integer, parameter :: noNonzeros = 5*(M-2)*(M-2) + 16*(M-2) + 12

  ! Coefficient matrix A in sparse CSR format
  type(CSR) :: A   
  
  ! right hand side
  real(kind=8), dimension(noUnknowns) :: b  

  ! solution vector
  real(kind=8), dimension(noUnknowns) :: u

  !------ Starting the code execution -------!

  ! Initialise the CSR matrix A and give it some memory
  Call setup_CSR(A, noUnknowns, noUnknowns, noNonzeros)

  ! initialise boundary condition
  b = 0.0d0
  
  
  ! In the example we will populate the matrix by hand
  ! stepping along each row of the u-field in x-direction
  ! column by column

  index=1
  do j = 1, M  ! looping over rows of the u-field 
     y = real(j, kind=8) / real(M+1, kind=8)

     do i = 1, M  ! looping over columns of the u-field
        x = real(i, kind=8) / real(M+1, kind=8)

        ! converting two-dimensional indices into 1 dimensional index
        RC = i + (j-1)*M

        ! Lookup table for north, east, south and west neighbour
        RN = RC + M
        RE = RC + 1
        RS = RC - M
        RW = RC - 1

        ! coefficient values
        coeffC = 4.0d0 * real( (M+1)*(M+1), kind=8)
        coeffN = - real( (M+1)*(M+1), kind=8)
        coeffE = - real( (M+1)*(M+1), kind=8)
        coeffS = - real( (M+1)*(M+1), kind=8)
        coeffW = - real( (M+1)*(M+1), kind=8)

        ! Set the row-start
        A%rowStart(RC) = index

        ! Now insert coefficients into the matrix, 
        ! modifying the allocation for boundary terms
        ! 
        ! There are nine different cases corresponding to
        ! the interior, the sides and corners of the u-field

        ! Interior point
        If ((i .gt. 1) .and. (i .lt. M) .and. (j .gt. 1) .and. (j .lt. M)) then
           A%colIndex(index) = RS
           A%value(index) = coeffS
           index = index + 1

           A%colIndex(index) = RW
           A%value(index) = coeffW
           index = index + 1

           A%colIndex(index) = RC
           A%value(index) = coeffC
           index = index + 1

           A%colIndex(index) = RE
           A%value(index) = coeffE
           index = index + 1

           A%colIndex(index) = RN
           A%value(index) = coeffN
           index = index + 1

        ! West boundary
        elseif ((i .eq. 1) .and. (j .gt. 1) .and. (j .lt. M)) then
           A%colIndex(index) = RS
           A%value(index) = coeffS
           index = index + 1

           A%colIndex(index) = RC
           A%value(index) = coeffC
           index = index + 1

           A%colIndex(index) = RE
           A%value(index) = coeffE
           index = index + 1

           A%colIndex(index) = RN
           A%value(index) = coeffN
           index = index + 1

        ! East boundary
        elseif ((i .eq. M) .and. (j .gt. 1) .and. (j .lt. M)) then
           A%colIndex(index) = RS
           A%value(index) = coeffS
           index = index + 1

           A%colIndex(index) = RW
           A%value(index) = coeffW
           index = index + 1

           A%colIndex(index) = RC
           A%value(index) = coeffC
           index = index + 1

           A%colIndex(index) = RN
           A%value(index) = coeffN
           index = index + 1

           ! we need to add the polution source to the bounday condition
           If ( (y .gt. y1) .and. ( y .lt. y2 ) ) then
              b(RC) = - coeffE * k * (( y2-y)*(y-y1))**p
           endif

        ! south bounday
        Elseif ((i .gt. 1) .and. (i .lt. M) .and. (j .eq. 1) ) then
           A%colIndex(index) = RW
           A%value(index) = coeffW
           index = index + 1

           A%colIndex(index) = RC
           A%value(index) = coeffC
           index = index + 1

           A%colIndex(index) = RE
           A%value(index) = coeffE
           index = index + 1

           A%colIndex(index) = RN
           A%value(index) = coeffN
           index = index + 1

        ! north boundary
        Elseif ((i .gt. 1) .and. (i .lt. M) .and. (j .eq. M)) then
           A%colIndex(index) = RS
           A%value(index) = coeffS
           index = index + 1

           A%colIndex(index) = RW
           A%value(index) = coeffW
           index = index + 1

           A%colIndex(index) = RC
           A%value(index) = coeffC
           index = index + 1

           A%colIndex(index) = RE
           A%value(index) = coeffE
           index = index + 1

        ! south west corner
        elseif ((i .eq. 1) .and. (j .eq. 1) ) then
           A%colIndex(index) = RC
           A%value(index) = coeffC
           index = index + 1

           A%colIndex(index) = RE
           A%value(index) = coeffE
           index = index + 1

           A%colIndex(index) = RN
           A%value(index) = coeffN
           index = index + 1

        ! south-east corner
        elseif ((i .eq. M) .and. (j .eq. 1)) then
           A%colIndex(index) = RW
           A%value(index) = coeffW
           index = index + 1

           A%colIndex(index) = RC
           A%value(index) = coeffC
           index = index + 1

           A%colIndex(index) = RN
           A%value(index) = coeffN
           index = index + 1

        ! north-west corner
        elseif ((i .eq. 1) .and. (j .eq. M)) then
           A%colIndex(index) = RS
           A%value(index) = coeffS
           index = index + 1

           A%colIndex(index) = RC
           A%value(index) = coeffC
           index = index + 1

           A%colIndex(index) = RE
           A%value(index) = coeffE
           index = index + 1

        ! North-east corner
        elseif ((i .eq. M) .and. (j .eq. M)) then
           A%colIndex(index) = RS
           A%value(index) = coeffS
           index = index + 1

           A%colIndex(index) = RW
           A%value(index) = coeffW
           index = index + 1

           A%colIndex(index) = RC
           A%value(index) = coeffC
           index = index + 1

        else
           print * ,"This should not be happen"
           print * , "Error in building matrix", i , j
           stop
        endif
     end do
  end do

  ! Check the correct number of elements have been allocated
  if ( index .eq. noNonzeros+1) then
     A%rowStart(A%nrow+1) = index
  else
     print *, "index = ", index, "  nnz = ", noNonzeros
     stop
  end if

  ! set initial solution to zero
  u =0.0d0

  ! Solve the system

  Call print_CSR(A)

  print *, 'Calling lincg_CSR'

  Call lincg_CSR(A, b, u, tol, maxIter, usedIter)

  print *, 'Number of iterations = ', usedIter

  ! Writing solution to disk

  print *, 'Writing solution to pollution.dat'

201 Format(f10.6,6x)
  Open(11, file="pollution.dat", status="unknown")
  Do j = 0, M+1
     y = real(j, kind=8) / real(M+1, kind=8)
     Do i = 0, M+1
        x = real(i, kind=8) / real(M+1, kind = 8)
        
        ! if interior point, write u-fields
        if ( (i .gt. 0) .and. (i .lt. M+1) .and. (j .gt. 0) .and. (j .lt. M+1) ) then
           RC = i + (j-1)*M
           write(11,201) u(RC)

        ! east boundary put the source
        elseif ( i .eq. M+1) then 
           if ((y .gt. y1) .and. ( y .lt. y2)) then
              write(11,201) k*((y-y2)*(y-y1))**p
           else
              write(11,201) 0.0d0
           endif
        else
           write(11,201) 0.0d0
        end if
     end Do
  end Do
  close(11)

  print *, 'Writing associated field file to pollution.general'

  !Writing an AVS field file
203 Format(80A)
  Open(12,file="pollution.general",status="unknown")
  write(12,203) "file = pollution.dat"
  write(12,*)"grid = ", M+2, " x ", M+2
  write(12,203) "format = ascii"
  write(12,203)"interleaving = record"
  write(12,203)"majority = row"
  write(12,203)"field = field0"
  write(12,203)"structure = scalar"
  write(12,203)"type = float"
  write(12,203)"dependency = positions"
  write(12,203)"positions = regular, regular, 0, 1, 0, 1"
  write(12,203)""
  write(12,203)"end"
  Close(12)


  ! Free the memory associated with A
  Call deallocate_csr(A)
  
end Program PollutionModel

File: pna/sparse/C/pollutionModel.c
#include "matrix.h"

int main()
{

/* set some parameters */
  const int M = 7; /* grid resultion */
  const int maxIter = 5000;
  const double tol = 1.0e-6;
  /* boundary conditions */
  const double y1 = 0.4;
  const double y2 = 0.6;
  const int p = 2;
  double k;

  /* indices and counters */
  int usedIter;
  int index; /* index of current element */
  int i; /* current i coordinate */
  int j; /* current j coordinate */

  double x,y; /* x and y at current coordinate */
  
  /* temporary store for coefficient data */
  double coeffC, coeffN, coeffE, coeffS, coeffW;

  /* local indices on interior mesh*/ 
  int RC, RN, RE, RS, RW;

  /* The linear sytem */
  int noUnknowns;
  int noNonZeros;

  /* The coefficient matrix */
  CSRmatrix A;

  /* pointers to the vectors */
  double *b, *u;

  FILE *fptr;

  /* starting the execution */

  /* calculate some derived parameters */
  /* k normalises the height of the chimney to 1 */
  k=pow( (2.0/(y2-y1)),(2*p) );
  /* the system size */ 
  noUnknowns = M*M;
  noNonZeros = 5*(M-2)*(M-2) + 16*(M-2) + 12;

  /* allocate the data objects */
  setup_CSR(&A, noUnknowns, noUnknowns, noNonZeros);
  b = (double *)malloc( (sizeof(double))*noUnknowns);
  u = (double *)malloc( (sizeof(double))*noUnknowns);


  /* set the boundary conditions */
  coeffC = 4.0 * (double)(M+1)*(M+1);
  coeffN=coeffE=coeffS=coeffW=-(double)(M+1)*(M+1);

  for(index=0;index<noUnknowns;index++){
    b[index]=0.0;  /* solving a homogenous eqn, i.e. RHS=0 */
                   /* chimney is added by hand later */
    u[index]=0.0;  /* initial trial solution is zero */ 
  }

  /* 
   * Populate the matrix by hand 
   * stepping along each row of the u-field in the x-direction
   * column by column 
   */

  index=0;

  for(j=0;j<M;j++){ /* looping over the rows if the u-field */
    y=(double)(j+1)/(double)(M+1);

    for(i=0;i<M;i++){ /* looping over the rows if the u-field */
      x=(double)(i+1)/(double)(M+1);
 
      /* converting the two-dimensional indices into 
       * one dimensional index */ 
      RC = i + j*M;

      /* Lookup table for ordinal neighbours */
      RN=RC+M;
      RE=RC+1;
      RS=RC-M;
      RW=RC-1;
      /* set the row-start */
      A.rowStart[RC]=index;
      
      /*
       * Now insert coefficients into the matrix, 
       * modifying the allocation for boundary terms
       * 
       * There are nine different cases corresponding to
       * the interior, the sides and corners of the u-field
       */

      /* interior */
      if( (i>0) && (i<(M-1)) && (j>0) && (j<(M-1)) ){
	A.colIndex[index] = RS;
	A.value[index]=coeffS;
	index++;

	A.colIndex[index] = RW;
	A.value[index]=coeffW;
	index++;

	A.colIndex[index] = RC;
	A.value[index]=coeffC;
	index++;

	A.colIndex[index] = RE;
	A.value[index]=coeffE;
	index++;

	A.colIndex[index] = RN;
	A.value[index]=coeffN;
	index++;
      }
      else if( (i==0) && (j>0) && (j<(M-1))){ /* west boundary */
	A.colIndex[index] = RS;
	A.value[index]=coeffS;
	index++;

	A.colIndex[index] = RC;
	A.value[index]=coeffC;
	index++;

	A.colIndex[index] = RE;
	A.value[index]=coeffE;
	index++;

	A.colIndex[index] = RN;
	A.value[index]=coeffN;
	index++;
      }
      else if( (i==(M-1)) && (j>0) && (j<(M-1)) ){ /* East boundary */

	A.colIndex[index] = RS;
	A.value[index]=coeffS;
	index++;

	A.colIndex[index] = RW;
	A.value[index]=coeffW;
	index++;

	A.colIndex[index] = RC;
	A.value[index]=coeffC;
	index++;

	A.colIndex[index] = RN;
	A.value[index]=coeffN;
	index++;

	/* if we need to add the pollution source to the boundary condition */
	if( (y>y1) && (y<y2)){
	  b[RC]= -coeffE * k * pow( ( (y2-y)*(y-y1) ), p);
	}

      }
      else if( (i>0) && (i<(M-1)) && (j==0) ){ /* South boundary */

	A.colIndex[index] = RW;
	A.value[index]=coeffW;
	index++;

	A.colIndex[index] = RC;
	A.value[index]=coeffC;
	index++;

	A.colIndex[index] = RE;
	A.value[index]=coeffE;
	index++;

	A.colIndex[index] = RN;
	A.value[index]=coeffN;
	index++;
      }
      else if( (i>0) && (i<(M-1)) && (j==(M-1)) ){ /* north boundary */
	A.colIndex[index] = RS;
	A.value[index]=coeffS;
	index++;

	A.colIndex[index] = RW;
	A.value[index]=coeffW;
	index++;

	A.colIndex[index] = RC;
	A.value[index]=coeffC;
	index++;

	A.colIndex[index] = RE;
	A.value[index]=coeffE;
	index++;

      }
      else if( (i==0) && (j==0) ){ /* SW corner */
	A.colIndex[index] = RC;
	A.value[index]=coeffC;
	index++;

	A.colIndex[index] = RE;
	A.value[index]=coeffE;
	index++;

	A.colIndex[index] = RN;
	A.value[index]=coeffN;
	index++;
      }
      else if( (i==(M-1)) && (j==0) ){ /* SE corner */
	A.colIndex[index] = RW;
	A.value[index]=coeffW;
	index++;

	A.colIndex[index] = RC;
	A.value[index]=coeffC;
	index++;

	A.colIndex[index] = RN;
	A.value[index]=coeffN;
	index++;
      }
      else if( (i==0) && (j==(M-1)) ){ /* NW corner */
	A.colIndex[index] = RS;
	A.value[index]=coeffS;
	index++;

	A.colIndex[index] = RC;
	A.value[index]=coeffC;
	index++;

	A.colIndex[index] = RE;
	A.value[index]=coeffE;
	index++;
      }
      else if( (i==(M-1)) && (j==(M-1)) ){ /* NE corner */
	A.colIndex[index] = RS;
	A.value[index]=coeffS;
	index++;

	A.colIndex[index] = RW;
	A.value[index]=coeffW;
	index++;

	A.colIndex[index] = RC;
	A.value[index]=coeffC;
	index++;
      }
      else{
	fprintf(stderr,"pollution:Error:Building matrix out of bounds [%d][%d]\n",i,j);
	return 1;
      }

    } /* end loop over i */
  } /* end loop over j */

  /* check the correct number of elements have been allocated */
  if(index==noNonZeros){
    A.rowStart[A.nrow]=index;
  }
  else{
    fprintf(stderr,"pollution:Error:wrong number of elements assigned index=%d nnz=%d\n",index,noNonZeros);
    return 1;
  }

  print_CSR(&A);
  
  linearCG_CSR(&A,b,u,tol,maxIter,&usedIter);

  printf("%s:Solution found in %i iterations\n",__FUNCTION__,usedIter);
  printf("%s:Writing solution to pollution.dat\n",__FUNCTION__);

  fptr=fopen("pollution.dat","w");
  for(j=0;j<=M+1;j++){
    y=(double)j/(double)(M+1);
    for(i=0;i<=M+1;i++){
      x=(double)i/(double)(M+1);
      /* If interior point write u-fields */
      if( (i>0) && (i<M+1) && (j>0) && (j<M+1) ){
	RC=i+(j-1)*M-1;
	  fprintf(fptr,"%10.6f\n",u[RC]);
      }
      else if( i==M+1 ){
	/* east boundary put the source */
	if( (y>y1) && (y<y2) ){
      	  fprintf(fptr,"%10.6f\n",k*pow( (((y-y2)*(y-y1))) , p )); 
	} 
	else{ 
      	  fprintf(fptr,"%10.6f\n",0.0); 
	} 
      } 
      else{ 
	fprintf(fptr,"%10.6f\n",0.0); 
      }
    }
  }
  fclose(fptr);

  printf("%s:Writing associated field file to polluton.general\n",__FUNCTION__);
  fptr=fopen("pollution.general","w");
  fprintf(fptr,"file = pollution.dat\n");
  fprintf(fptr,"grid = %d x %d\n",M+2,M+2);
  fprintf(fptr,"format = ascii\n");
  fprintf(fptr,"interleaving = record\n");
  fprintf(fptr,"majority = row\n");
  fprintf(fptr,"field = field0\n");
  fprintf(fptr,"structure = scalar\n");
  fprintf(fptr,"type = float\n");
  fprintf(fptr,"dependency = positions\n");
  fprintf(fptr,"positions = regular, regular, 0, 1, 0, 1\n");
  fprintf(fptr,"\nend\n");
  fclose(fptr);



  return 0;
}
